### ISSUE TITLE
More distributed and resilient transactions

### ISSUE THREAD

**Issue by kshlm** [gluster/glusterd2#919]
The transaction framework as it exists right now, works well. But it has a couple of drawbacks owing to it having a leader who drives the transactions steps.

As we depend on the leader to drive transactions, we also depend on the leader to recover from failures. In case the leader of a transaction fails (quits/crashes) in the middle of a transaction, the transaction would remain unfinished and any already performed steps would be undone. This defeats one of the reasons for creating the new transaction framework (undo support).

To address the above, I'm proposing the following changes to the transaction framework, to make it mode distributed, and allow each peer to act independently.

## Proposed transaction flow

Initially a transaction still starts from a single peer. This peer receives the ReST request, and performs the necessary actions to generate the list of steps required for the transaction.

The peer will add this list of steps to a new transaction namespace (say NewTxnNS). The peer will then watch the transaction in the NewTxnNS and wait for it to pass/fail/complete.

Every peer in the cluster, keeps watching on the NewTxnNS for new transactions. When a new transaction is added, they fetch the list of steps from the store and execute it independently. Each peer also updates the store of the success or failure of its execution.

The peer that received the request will wait for all required peers to update the store, and reply back to the client with the result.

In case of a failure, a peer would automatically undo the steps it had performed, and update the store with the failure. The remaining peers in the transaction keep a watch for any failures, and independently undo the steps they performed.

### Things to think about

Changing to the new design has a couple of issues that need to be solved.

#### Synchronizing steps / Step dependencies

The new design moves the transaction from a synchronous model to an asynchronous model. But our transactions have steps which depend on previous steps, possibly on another peer to be completed before it can be executed.

For eg., the commit steps in the existing transactions can only be performed after all previous steps have been completed successfully.

#### Modifying global data structures

_Global data structures include volinfo, snapinfo excluding node specific data structs like brickinfo, daemon info etc._

The current design does this by having the transaction leader perform the actual modifications on the global data structures, and avoiding conflicts by taking locks on the structures.

With the new design, we can easily elect a leader (for a single transaction, or to keep it simpler, globally for all transactions), and have it perform the locks and modifications. If an elected leader goes down, another leader will be elected and will take over operations.

But, this requires that we solve the problem of locking. Locking is currently performed by taking etcd locks, which are not transferable to another etcd client. This would prevent the next leader from taking ownership of the locks and performing the required modifications.

## Implementation

The above design can be implemented largely without affecting the consumers of the current transaction framework. The exposed API and the workflow would still remain the same. All required changed can be done internally within the transaction package.

The framework consumers may need to do small updates to their code to address changes required for step synchronization/ step dependencies.

**Comment by atinmu** [comment]
Putting down my immediate thoughts:

We need to figure out a way to see how the step dependencies and synchronization between different phases can be handled with this model before getting into the implementation part as if we can’t solve this part the whole purpose of this enhancement will be defeated.

Also does this also mean every transactions/API/CLI calls will become asyncronous with this model? If so don’t we anticipate any changes in API/CLI especially with capabilities of querying for the status of individual commands?

**Comment by shtripat** [comment]
My 2 cents, regarding locks, can we think of maintaining global lock details in a namespace which all the peers have access to and use this in addition while acquiring lock if there is a re-election of leader?

**Comment by kshlm** [comment]
> We need to figure out a way to see how the step dependencies and synchronization between different phases can be handled with this model before getting into the implementation part as if we can’t solve this part the whole purpose of this enhancement will be defeated.

I have a very basic solution that would do this. We could have a sync step, that can be added between 2 steps that need to be synchronized. The peers would pause at this step, wait for all other peers to execute this particular step, before continuing with the next step. This can ensure that all previous steps are done everywhere before the next step is executed.

Implementing this can also be simple.
- All peers executing the step are given an  etcd namespace to watch on.
- When the sync step is executed, the peers add themselves (a key) into the namespace
- They then wait till the namespace has enough keys added (number of keys added = number of peers in transaction)

This is a really simple way, but doesn't give you lots of granularity. This will cause all peers to wait at the step even though the next steps may not be executing on them.



> Also does this also mean every transactions/API/CLI calls will become asyncronous with this model? If so don’t we anticipate any changes in API/CLI especially with capabilities of querying for the status of individual commands?

As I'd mentioned in the descriptions and as @prashanthpai said, the API will continue to remain the same. The visible behaviour to the users of the framework will not change.

**Comment by kshlm** [comment]
> My 2 cents, regarding locks, can we think of maintaining global lock details in a namespace which all the peers have access to and use this in addition while acquiring lock if there is a re-election of leader?

This is most likely what we'll need to do.

**Comment by aravindavk** [comment]
What will happen to ongoing request if initiator node goes down. Will application gets connection refused or all the APIs need to be changed as async APIs?

**Comment by kshlm** [comment]
> What will happen to ongoing request if initiator node goes down. Will application gets connection refused or all the APIs need to be changed as async APIs?

The client gets a disconnect, and will not know about the status of the request. But if the transaction was initiated, it would continue independently.

**Comment by aravindavk** [comment]
> The client gets a disconnect, and will not know about the status of the request. But if the transaction was initiated, it would continue independently.

Looks good in a way(wrt to Transaction completeness and Rollback). But state mismatch between Application and Glusterd2, Application thinks request is failed but it may be successful.  If APIs are async, applications will have reference number to check back the status of the request. Any plans to address this issue?

**Comment by aravindavk** [comment]
If Client disconnect can be detected or when initiator node goes down, how about initiating rollback instead of completing the request.

**Comment by aravindavk** [comment]
Other problems:

- If initiator node goes down between transaction, Rollback steps from that node will not be executed. Need a mechanism to call rollback once that node comes back up.
- Repeated requests. This is not a problem in most cases(For example Volume start API called twice). But in some APIs which creates new entity and assigns ID Like Dynamic Volume provision, Volume name is optional and no way to differentiate the repeat request. (We can change this behavior back to making Volume name as required field, so that duplicate request can be identified by Volume name)

**Comment by kshlm** [comment]
> If initiator node goes down between transaction, Rollback steps from that node will not be executed. Need a mechanism to call rollback once that node comes back up.

This is required not just for the initiator node, but for any involved node going down. Handling it for the larger case will handle it for this specific case.

> Repeated requests. This is not a problem in most cases(For example Volume start API called twice). But in some APIs which creates new entity and assigns ID Like Dynamic Volume provision, Volume name is optional and no way to differentiate the repeat request. (We can change this behavior back to making Volume name as required field, so that duplicate request can be identified by Volume name)

I agree  that this is a problem specifically with IVP. But for other cases, I'd say we need to ask that clients don't try to maintain state, and that we should ensure that our APIs are idempotent. This way clients should be able to retry any request, and get back the result they originally expected.

**Comment by aravindavk** [comment]
> I agree that this is a problem specifically with IVP. But for other cases, I'd say we need to ask that clients don't try to maintain state, and that we should ensure that our APIs are idempotent. This way clients should be able to retry any request, and get back the result they originally expected.

+1

I will revert that patch and make Volume name as required again. Just to make sure, I will look into all the existing APIs to see this problem exists.

**Comment by prashanthpai** [comment]
Some thoughts on the topic idempotency and txn leader :

Just clarifying terms used to developers who don't work on glusterd2:

*initiator node*: the node that received the HTTP request from a client for a management operation (volume create, start etc)

*distributed lock*: taken by a node using the [etcd API](https://godoc.org/github.com/coreos/etcd/clientv3/concurrency#Mutex.Lock). When the lock owner goes down, etcd frees the lock. These locks are co-operative and advisory in nature.

*local lock*: a lock taken within the glusterd2 process, likely to be a [mutex](https://github.com/gluster/glusterd2/blob/master/pkg/utils/mutex.go).

For the first iteration of improvements: I'm proposing to not have a separate elected leader other than initiator node. The initiator node takes the lock. In the event of initiator node going down and lock being released, there's no further steps (Do functions) to be executed. Only rollback/undo has to be done. Other nodes part of the txn can run undo functions when they detect that the initiator node has gone down. Running these undo doesn't have to take a distributed lock (rather a local node level lock) if we ensure certain rules are in place.

For this to work, we have to follow a model which most traditional database systems use - a journal (temporary state) + on-disk reality (permanent state). For us, the etcd data namespace is the *permanent state* and the etcd txn namespace is the *temporary state*. We enforce following rules: 

* Changing the data namespace must only be done once at the end of transaction and only by initiator node. Intermediate steps and non-initiator nodes must not modify the data namespace. The initiator node thus becomes the natural "leader" authorised to modify locked resource in data namespace.
* `Undo` must not revert things in data namespace (on leader node or otherwise). IOW, `Undo` is allowed only to modify things in its txn namespace and things only local to that node (like files/directories/devices).
* `Undo` functions from previous unfinished txns are run when a failed node restarts. This node doesn't accept/process further conflicting requests until it has caught up with running all pending rollbacks. Running `Undo` is to be carried out taking a node-level lock.

**Cons**:
Between the events of "commit to etcd data namespace" and "replying to client", two common failures can happen: initiator node can go down or the client can hang up. As per above rules, that initiator node can only undo local operations, not the permanent commit. In that case, the client cannot retry the same request. Generally (but not necessarily), requests that create new named (by client) resource (such as volume) using a `POST` request doesn't have to idempotent. Creating a named (by server) resource (such as volume) can be idempotent - the client can retry that.

In a nutshell - we do not give an option to undo a permanent commit to data namespace.

Clearly, there are shortcomings with this approach. I'm okay to live with that for the first wave of proposed improvements.

**Comment by atinmu** [comment]
Getting the design ready and a first cut implementation of this enhancement is the target of GCS-Sprint1. Taking this enhancement to the completion will happen over multiple sprints.

**Comment by phlogistonjohn** [comment]
Here are some of my initial reactions to the original post. I realize some of this has already been discussed in subsequent comments after skimiming them but instead of editing and re-editing my notes I'm just going to drop them here and then react to comments individually. Please feel free to point out anything in my reaction that has already been settled or modified in case I missed it.

> The peer will add this list of steps to a new transaction namespace (say NewTxnNS). The peer will then watch the transaction in the NewTxnNS and wait for it to pass/fail/complete.

Might I propose a term like "pending" or "in progress" rather than new. New seems like an odd term for this namespace to me.


> Every peer in the cluster, keeps watching on the NewTxnNS for new transactions. When a new transaction is added, they fetch the list of steps from the store and execute it independently. Each peer also updates the store of the success or failure of its execution.

> The new design moves the transaction from a synchronous model to an asynchronous model. But our transactions have steps which depend on previous steps, possibly on another peer to be completed before it can be executed.

Can we go a little more in depth as to how coordination around updating the store for each transation step would go?
I imagine a structure like:
{namespace}/{txnid}/results/{peerid}/{stepid}
where each peer can independently update its own stuff @ {peerid}/{stepid}
but other nodes can query and/or watch those results.


> The peer that received the request will wait for all required peers to update the store, and reply back to the client with the result.

That sounds good but I want to dig a little deeper into how the finalization of the txn happens.
I think that we want to make updating any parts of the etcd store happens as part of a final step in
the transaction and is not performed "outside" of the transaction. Thus in the case that the initiating peer dies all that is lost is a response to the client.

This does imply that the transaction itself needs to be removed from the store as a seperate "process" from responding to the client of the transaction's success. Would it be possible to put an expiration on a completed transaction in etcd in case the initiating peer dies?

> With the new design, we can easily elect a leader (for a single transaction, or to keep it simpler, globally for all transactions), and have it perform the locks and modifications. If an elected leader goes down, another leader will be elected and will take over operations.

Similarly, to what I wrote above, perhaps transactions need a "step 0" to set up entries in the store
and a "step X" at the end for finalization. This would generally only be performed by on peer. Intially, these steps might be implicit and always performed by the initiating peer, however over time I think it would become explicit. Plus no locks would persist outside of a single transaction step.

This doesn't solve the locking issue but does make it clear that transaction's come first before any of the associated modifications to the db they would make. I think this is especially relevant given a more complex operation on the store like IVP or node migration where many items in the store may be involved.

**Comment by phlogistonjohn** [comment]
> I agree that this is a problem specifically with IVP. But for other cases, I'd say we need to ask that clients don't try to maintain state, and that we should ensure that our APIs are idempotent. This way clients should be able to retry any request, and get back the result they originally expected.

> I will revert that patch and make Volume name as required again. Just to make sure, I will look into all the existing APIs to see this problem exists.

This is where an async api would be useful IMO. The server would reply back to a request with an immediate response of a uniqe "cookie" that the client then uses to track the status of the transaction itself rather than relying on a particular tcp connection to live. That said I don't think the issue is a dealbreaker in a sync api. You might just change the api such that a "cookie" is required at create time and so that if you come back the api with the same cookie the server knows what transaction it match it with. (That said I still personally prefer the async model)

There's also an aspect of this that we've seen in heketi land that is not solved (by heketi's approach to) the async api. This is when something like the k8s clients constantly retry requests and in this case I want to hear from @raghavendra-talur  and @Madhu-1  as they started on a solution for this in Heketi land.

**Comment by phlogistonjohn** [comment]
> Can we start writing the design pieces into a google document where all can review and provide comments? Ideally what I'm looking at is a design document which once all the comments are addressed and everyone is in agreement, can get to the wiki and replace our existing transaction framework.

I'd strongly suggest that the design doc live as a file within the glusterd2 repo. You get the benefit of using the PR system for review and feedback and when all is said and done it's a file in your repo that can live along side the code that implements it. Future PRs can update the doc. IMO google docs are not nearly as transparent and discover-able and wiki pages often quickly become neglected and out of date (plus the lack of a PR-style feedback process).

We have been using this approach for heketi and it's worked well so far.

**Comment by phlogistonjohn** [comment]
> For this to work, we have to follow a model which most traditional database systems use - a journal (temporary state) + on-disk reality (permanent state). For us, the etcd data namespace is the permanent state and the etcd txn namespace is the temporary state. We enforce following rules:

+1 to the general gist of a journal.

>    Changing the data namespace must only be done once at the end of transaction and only by initiator node. Intermediate steps and non-initiator nodes must not modify the data namespace. The initiator node thus becomes the natural "leader" authorised to modify locked resource in data namespace.

This works really well for transactions that update a single "object" in the store. I think we need to be a little careful about dealing with scenarios that either need to draw data from the underlying system, like say pulling in the output of a lvm command or updating many objects.  (More on this below)

>    Undo must not revert things in data namespace (on leader node or otherwise). IOW, Undo is allowed only to modify things in its txn namespace and things only local to that node (like files/directories/devices).

+1 

>    Undo functions from previous unfinished txns are run when a failed node restarts. This node doesn't accept/process further conflicting requests until it has caught up with running all pending rollbacks. Running Undo is to be carried out taking a node-level lock.

+1 this is one of the parts of the story that is missing from heketi when we added our version of "journaling". The plus side is we know what went wrong but without post-restart undo a human has to figure out how to clean up the system. Even with external tooling this is frustrating when in theory the server could have done it for you

I think that journal may need a temporary data namespace in addition to a pure transaction namespace (note: this doesn't neccessarily mean a real etcd namespace it could just as well be some sort of subtree).
For example, consider the following IVP style scenario: a user sends a request to gd2 requesting a 4GiB volume, using replica-3 and arbiter enabled, named "bob". In order to create the volume we first need to figure out where the new bricks will be located by examining the system for where we can put two bricks of size X and one of size Y. We need to do this in a way that other incoming transactions will not change the devices under consideration and such that if we can satisfy the request without error that space is then set aside for the remaining steps of the transaction to use (or be undone in case of a failure).

One way I can imagine doing this is that in the etcd version of a db transaction a single gd2 transaction step finds the devices and creates brick objects in the temporary namespace. The act of finding space for these bricks takes into account the fully created bricks in the data namespace as well as the not-yet-complete bricks in the temporary namespace.
(In the heketi this would be equivalent to the "build" step of an operation, which is done in a single boltdb transaction). Once this step is done other steps can satisfy the individual changes needed on each peer and then the "step X" that I mentioned above would finalize it all in the permanent data area.

**Comment by phlogistonjohn** [comment]
I think one thing that can help the design is if the doc contains a few example scenarios and how the system would handle them.

Personally, I'd like to cover how it might handle stuff going from low complexity to high:
* Simple volume creation - happy path
* Simple volume creation - worker node dies
* Simple volume creation - initiating node dies
* Rough sketch IVP - happy path  (for cover multi object coverage)
* Rough sketch IVP - node(s) die
* Rough sketch node replace
* Rough sketch node replace - node(s) die

I included the last two in particular because these are very flaky even in current heketi and if the design around them seems solid we should have confidence in the overall approach. 
(Note: I'm not trying to design IVP & device management stuff here, just how the framework would approach complex multi-step changes to many items type stuff - the actual data structures and step can and should be hand-wavy)

**Comment by kshlm** [comment]
I'v posted the doc for review at #1003. [Direct link](https://github.com/kshlm/glusterd2/blob/new-txn-fw-doc/doc/transaction.md) for easy viewing.

**Comment by atinmu** [comment]
I'm putting this back to GCS-blocker as handling the rollback while the originator glusterd goes down look very important.
